### PR TITLE
Creates PowerShell Samples / Tests

### PR DESCRIPTION
**Pull request by VasuBhog**
Created the samples and tests following C# / Javascript samples.

The PowerShell integration issues will be tracked here:
https://github.com/Azure/azure-functions-sql-extension/issues/448

Had to update the local.settings.json in order to pass the ad hoc build.

Other known issues found while developing:
If a test fails, the function host won't start as the port is unavailable and therefore needs to kill the port to continue some of the tests. [(Plan to create a fix in another PR)](https://github.com/Azure/azure-functions-sql-extension/issues/442).
PowerShell worker issue: https://github.com/Azure/azure-functions-sql-extension/issues/443

### CODE REVIEW

**Comment by VasuBhog** [comment]
AddProductWithIdentity seems to be failing at random as it passes in some of the ad-hoc builds but then a different `AddProductWithIdentity` test fails in PR Validation. 

I will be marking all of them as unstable for now, and track [here](https://github.com/Azure/azure-functions-sql-extension/issues/448)

File: samples/samples-powershell/.eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (Charles-Gagnon) [inline samples/samples-powershell/.eslintrc.json:1]: Why is there an eslint file in the powershell dir?

File: samples/samples-powershell/.funcignore [hunk 2]
```diff
@@ -0,0 +1,6 @@
+*.js.map
```
Comment (Charles-Gagnon) [inline samples/samples-powershell/.funcignore:1]: Clean up this file to only include the necessary stuff, js and ts for example definitely don't need to be in here

File: samples/samples-powershell/.gitignore [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+npm-debug.log*
+yarn-debug.log*
+yarn-error.log*
+lerna-debug.log*
+
+# Diagnostic reports (https://nodejs.org/api/report.html)
+report.[0-9]*.[0-9]*.[0-9]*.[0-9]*.json
+
+# Runtime data
+pids
```
Comment (Charles-Gagnon) [inline samples/samples-powershell/.gitignore:13]: You shouldn't need most of these - they'll be inherited from the ones higher up. Delete this and then just add things back one by one as needed

File: samples/samples-powershell/host.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  "logging": {
+    "applicationInsights": {
+      "samplingSettings": {
+        "isEnabled": true,
+        "excludedTypes": "Request"
+      }
+    }
+  },
+  "extensionBundle": {
+    "id": "Microsoft.Azure.Functions.ExtensionBundle.Preview",
```
Comment (Charles-Gagnon) [inline samples/samples-powershell/host.json:12]: Did you verify that the tests are using the build extension DLL and not the one from the bundle?

**Review by Charles-Gagnon — COMMENTED** [review]
1. Make sure you merge in latest from main to remove the serialization changes
2. If you haven't been able to verify that it's using the locally built extension DLL during tests (both locally and in the pipeline) then what would probably be best is to remove Powershell from the list of supported languages and then follow up with a PR that actually enables the tests (along with any changes necessary to run using the locally built DLL). This will make that simpler to review and get this in so that others can at least see the various samples and use them while giving you time to figure out the test situation so we're actually testing latest changes to the code

**Comment by Charles-Gagnon** [comment]
Also it would be good to add a .vscode folder with the recommended extensions for these samples (AF and Powershell extensions at least, not sure if there's any more that you've found helpful)

See https://github.com/Azure/azure-functions-sql-extension/pull/458/files#diff-5901d1ff3dfb2f6d0846ab6a4aeab48d916c3a07b3074de578cb760e3be9bb50R3 for an example

**Comment by chlafreniere** [comment]
How about including extension recommendations for the PS + Azure functions extensions?

**Comment by VasuBhog** [comment]
> 1. Make sure you merge in latest from main to remove the serialization changes
> 2. If you haven't been able to verify that it's using the locally built extension DLL during tests (both locally and in the pipeline) then what would probably be best is to remove Powershell from the list of supported languages and then follow up with a PR that actually enables the tests (along with any changes necessary to run using the locally built DLL). This will make that simpler to review and get this in so that others can at least see the various samples and use them while giving you time to figure out the test situation so we're actually testing latest changes to the code

@Charles-Gagnon, after discussion with @lucyzhang929 we found that the local and pipeline are using the latest preview extension bundle. 

Here are links to the latest pipeline runs that indicate we are using the correct latest version (4.3.1) of the Preview Extension bundle:
https://mssqltools.visualstudio.com/CrossPlatBuildScripts/_build/results?buildId=178320&view=logs&j=911909e3-f132-51e7-cb10-4a0c4ab1d143&t=c2542b5c-8cb4-5274-ab21-603e725c31e3&l=13289

https://mssqltools.visualstudio.com/CrossPlatBuildScripts/_build/results?buildId=178320&view=logs&j=911909e3-f132-51e7-cb10-4a0c4ab1d143&t=ad6b2e8f-d421-501d-db33-e4ace593a6cb&l=10

Lucy and I discussed steps locally that could be documented further to test locally built DLL:
1. run `dotnet build` in root of the repo
2. Copy the DLL in src/bin/Debug/netstandard2.0 to the extension bundle (samples you want to test - step 3).
3. Go to the directory of the extension bundle you want to change - for example azure-functions-sql-extension/samples/samples-powershell` and use `func GetExtensionBundlePath` to see where to copy the DLL found in step 2 into.
4. Copy the DLL from step 2 to the path found in step 3 and run `Dotnet test` at the root of the repo as it should use the latest builds then.

**Comment by Charles-Gagnon** [comment]
It's not the latest version of the bundle I'm worried about, it's the version of the extension DLL that is being loaded by the functions during tests. It doesn't help a lot to be using the one from the bundle since the tests should be verify that the code in the repo at the commit the test is running against still works as expected.

1. Have you verified that Lucy's change to copy the DLL to the bundle in the pipeline also works for the Powershell tests? https://github.com/Azure/azure-functions-sql-extension/blob/main/builds/azure-pipelines/template-steps-build-test.yml#L74 It should, but let's make 100% sure so at least we can rely on the pipeline to be accurate
2. Documenting manual steps to run the tests locally with the locally built DLL is fine for now but please make an issue for us to follow up with making that done automatically - possibly even embedded as a startup task in the tests themselves.

**Comment by VasuBhog** [comment]
> How about including extension recommendations for the PS + Azure functions extensions?

Pushed latest commit to include recommendations for PS and Az Func.

File: test/Integration/test-powershell/AddProductsNoPartialUpsert/run.ps1 [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+# Trigger binding data passed in via param block
+param($Request)
+$totalUpserts = 100;
+# Write to the Azure Functions log stream.
+Write-Host "[QueueTrigger]: $Get-Date starting execution $queueMessage. Rows to generate=$totalUpserts."
+
+# Update req_body with the body of the request
+# Note that this expects the body to be a JSON object or array of objects 
+# which have a property matching each of the columns in the table to upsert to.
+$start = Get-Date
```
[the remark this answers is not in the document]
Comment (VasuBhog) [reply test/Integration/test-powershell/AddProductsNoPartialUpsert/run.ps1]: no can remove as this followed the sample closely but for test it does not need it.

File: samples/samples-powershell/.gitignore [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# Logs
+logs
```
[the remark this answers is not in the document]
Comment (VasuBhog) [reply samples/samples-powershell/.gitignore]: removed files as I checked to see top-level root and what else was needed.

File: samples/samples-powershell/.gitignore [hunk 7]
```diff
@@ -1,13 +0,0 @@
...
-npm-debug.log*
-yarn-debug.log*
-yarn-error.log*
-lerna-debug.log*
-
-# Azure Functions artifacts
-bin
-obj
-appsettings.json
-local.settings.json
```
[the remark this answers is not in the document]
Comment (VasuBhog) [reply samples/samples-powershell/.gitignore]: Yes, all other languages (C#, JS, Python) have a .gitignore.

[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply samples/samples-powershell/.gitignore]: As mentioned in another comment - we should be generating a blank Powershell function project and copying over all the default files generated from that (after opening it in VS Code so it generates that stuff too)

**Comment by lucyzhang929** [comment]
Based on the logging from Azure Functions in the pipeline run Vasu linked above, it is loading the extension bundle from the same folder we copy the sql dll into. Although to be 100% sure, maybe we can add some temporary logging and ensure that the logs are being printed in the pipeline. 

Automatically copying over the latest dll to our local extension bundle would be great.

File: samples/samples-powershell/.gitignore (hunk 7 above)
[the remark this answers is not in the document]
Comment (VasuBhog) [reply samples/samples-powershell/.gitignore]: So creating a blank HTTP Trigger PowerShell function creates:
- function.json / run.ps1
- .funignore 
- .gitignore
- requirements.psd1
- profile.ps1

Just pushed these into commit which should be the "defaults" for our powershell samples folder.

File: samples/samples-powershell/host.json (hunk 4 above)
Comment (VasuBhog) [reply samples/samples-powershell/host.json:12]: Verified and see Lucy's comment.

**Review by Charles-Gagnon — APPROVED** [review]
Changes look good, but you have some test failures now. Please make sure those are addressed before merging in.

File: test/Integration/SqlOutputBindingIntegrationTests.cs [hunk 8]
```diff
@@ -46,6 +45,9 @@ public void AddProductTest(int id, string name, int cost, SupportedLanguages lan
         [SqlInlineData(1, "Test", 5)]
         [SqlInlineData(0, "", 0)]
         [SqlInlineData(-500, "ABCD", 580)]
+        // Currently PowerShell returns null if the output
```
Comment (lucyzhang929) [inline test/Integration/SqlOutputBindingIntegrationTests.cs:48]: Currently PowerShell returns null when the parameter for name is an empty string

File: samples/samples-powershell/AddProductWithIdentityColumn/run.ps1 [hunk 9]
```diff
@@ -0,0 +1,26 @@
+using namespace System.Net
+
+# Trigger binding data passed in via param block
+param($Request)
+
+# Write to the Azure Functions log stream.
+Write-Host "PowerShell function with SQL Output Binding processed a request."
+
+# Update req_body with the body of the request
```
Comment (lucyzhang929) [inline samples/samples-powershell/AddProductWithIdentityColumn/run.ps1:9]: Update comment - we are not using request body here.

File: samples/samples-powershell/AddProductWithIdentityColumnIncluded/run.ps1 [hunk 10]
```diff
@@ -0,0 +1,27 @@
+using namespace System.Net
+
+# Trigger binding data passed in via param block
+param($Request)
+
+# Write to the Azure Functions log stream.
+Write-Host "PowerShell function with SQL Output Binding processed a request."
+
+# Update req_body with the body of the request
```
Comment (lucyzhang929) [inline samples/samples-powershell/AddProductWithIdentityColumnIncluded/run.ps1:9]: Update comment. Several other functions have the same comment as well. Make sure to update or remove them.

**Review by chlafreniere — APPROVED** [review]

File: samples/samples-powershell/AddProductWithIdentityColumnIncluded/run.ps1 [hunk 11]
```diff
@@ -0,0 +1,25 @@
+using namespace System.Net
+
+# Trigger binding data passed in via param block
+param($Request)
+
+# Write to the Azure Functions log stream.
+Write-Host "PowerShell function with SQL Output Binding processed a request."
+
+# Update req_query with the body of the request
```
Comment (lucyzhang929) [inline samples/samples-powershell/AddProductWithIdentityColumnIncluded/run.ps1:9]: Update comment

File: samples/samples-powershell/AddProductWithMultiplePrimaryColumnsAndIdentity/run.ps1 [hunk 12]
```diff
@@ -0,0 +1,25 @@
+using namespace System.Net
+
+# Trigger binding data passed in via param block
+param($Request)
+
+# Write to the Azure Functions log stream.
+Write-Host "PowerShell function with SQL Output Binding processed a request."
+
+# Update req_query with the body of the request
```
Comment (lucyzhang929) [inline samples/samples-powershell/AddProductWithMultiplePrimaryColumnsAndIdentity/run.ps1:9]: Update comment

File: samples/samples-powershell/AddProductsWithIdentityColumnArray/run.ps1 [hunk 13]
```diff
@@ -0,0 +1,29 @@
+using namespace System.Net
+
+# Trigger binding data passed in via param block
+param($Request)
+
+# Write to the Azure Functions log stream.
+Write-Host "PowerShell function with SQL Output Binding processed a request."
+
+# Update req_body with the body of the request
```
Comment (lucyzhang929) [inline samples/samples-powershell/AddProductsWithIdentityColumnArray/run.ps1:9]: Update comment

File: test/Integration/test-powershell/AddProductExtraColumns/run.ps1 [hunk 14]
```diff
@@ -0,0 +1,26 @@
+using namespace System.Net
+
+# Trigger binding data passed in via param block
+param($Request)
+
+# Write to the Azure Functions log stream.
+Write-Host "PowerShell function with SQL Output Binding processed a request."
+
+# Update req_body with the body of the request
```
Comment (lucyzhang929) [inline test/Integration/test-powershell/AddProductExtraColumns/run.ps1:9]: Update comment

File: test/Integration/test-powershell/AddProductMissingColumns/run.ps1 [hunk 15]
```diff
@@ -0,0 +1,27 @@
+using namespace System.Net
+
+# Trigger binding data passed in via param block
+param($Request)
+
+# Write to the Azure Functions log stream.
+Write-Host "PowerShell function with SQL Output Binding processed a request."
+
+# Update req_body with the body of the request
```
Comment (lucyzhang929) [inline test/Integration/test-powershell/AddProductMissingColumns/run.ps1:9]: Update comment

File: test/Integration/test-powershell/AddProductMissingColumnsExceptionFunction/run.ps1 [hunk 16]
```diff
@@ -0,0 +1,27 @@
+using namespace System.Net
+
+# Trigger binding data passed in via param block
+param($Request)
+
+# Write to the Azure Functions log stream.
+Write-Host "PowerShell function with SQL Output Binding processed a request."
+
+# Update req_body with the body of the request
```
Comment (lucyzhang929) [inline test/Integration/test-powershell/AddProductMissingColumnsExceptionFunction/run.ps1:9]: Update comment

**Pull request merged**